### DISCUSSION TITLE
web gui, pkl, and yaml generation

### DISCUSSION THREAD

**Discussion by tannerhladek** [apple/pkl#172] [Q&A] [answered]
I am looking to build a web gui to enable visual yaml generation for a machine learning processe. Pkl's validation and yaml generation seem great, but I am curious if I am barking up the wrong tree. Thoughts and suggestions are much appreciated :)

**Comment by bioball** [comment]
Can you clarify your question? Do you want is a GUI that generates YAML? If so, I'm not sure why Pkl would be helpful here. Unless you wanted to embed a Pkl editor within your web GUI?

**Reply by tannerhladek** [reply]
Understood, the primary goal is to create a GUI for generating YAML. Considering the redundancy of using Pkl given that all necessary data is already passed to the backend, it might be more straightforward to directly generate YAML from user inputs in the GUI... Out of curiosity, has Pkl been utilized as a master template to ensure that the generated YAMLs adhere to a specific schema? This could clarify its role in the project.

**Reply by bioball** [reply, marked as answer]
Right now, Pkl isn't great at _validating_ external YAML. You might want to consider [CUE](https://cuelang.org/docs/usecases/validation/) for this, or [JSON Schema](https://json-schema.org). If you want to use JSON Schema, you can certainly use Pkl to generate that JSON Schema.

We do want to support the validation use-case at some point, but it's not there yet.

**Reply by tannerhladek** [reply]
Thanks for your insights on Pkl, CUE, and JSON Schema. I'll explore these options for my project and keep an eye on Pkl's evolution for future needs! Your guidance is much appreciated :)
